nixos/modules/services/web-servers/apache-httpd2/default.nix

```diff
@@ -0,0 +1,123 @@
+        default = "wwwrun";
+        description = ''
+          User account under which httpd children processes run.
+          If you require the main httpd process to run as
+          <literal>root</literal> add the following configuration:
+          <programlisting>
+          systemd.services.httpd.serviceConfig.User = lib.mkForce "root";
+          </programlisting>
+        '';
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "wwwrun";
+        description = ''
+          Group under which httpd children processes run.
+        '';
+      };
+
+      configuration = mkOption {
+        description = "Apache2 configuration";
+        type = with types;
+          let
+            self = 
+              oneOf [
+                (attrsOf (oneOf [
+                  str
+                  int
+                  (listOf (oneOf [ str int (listOf (oneOf [ str int ])) ]))
+                  (attrsOf self)
+                ]))
+                (listOf (oneOf [ str self]))
+              ];
+          in
+            self // { description = "loop breaker"; };
+      };
+    };
+  };
+
+  config =
+    let
+      httpdConf = pkgs.writeText "apache.cfg" (configParser cfg.configuration);
+    in 
+      (mkIf cfg.enable {
+        users.users = optionalAttrs (cfg.user == "wwwrun") {
+          wwwrun = {
+            group = cfg.group;
+            description = "Apache httpd user";
+            uid = config.ids.uids.wwwrun;
+          };
+        };
+
+        users.groups = optionalAttrs (cfg.group == "wwwrun") {
+          wwwrun.gid = config.ids.gids.wwwrun;
+        };
+
+        systemd.services.httpd2 = {
+          description = "Apache HTTPD";
+          wantedBy = [ "multi-user.target" ];
+          # wants = concatLists (map (certName: [ "acme-finished-${certName}.target" ]) dependentCertNames);
+          after = [ "network.target" ]; # ++ map (certName: "acme-selfsigned-${certName}.service") dependentCertNames;
+          # before = map (certName: "acme-${certName}.service") dependentCertNames;
+
+          path = [ pkg pkgs.coreutils pkgs.gnugrep ];
+
+          # environment =
+          #   optionalAttrs cfg.enablePHP { PHPRC = phpIni; }
+          #   // optionalAttrs cfg.enableMellon { LD_LIBRARY_PATH  = "${pkgs.xmlsec}/lib"; };
+
+          preStart =
+            ''
+                # Get rid of old semaphores.  These tend to accumulate across
+                # server restarts, eventually preventing it from restarting
+                # successfully.
+                for i in $(${pkgs.util-linux}/bin/ipcs -s | grep ' ${cfg.user} ' | cut -f2 -d ' '); do
+                  ${pkgs.util-linux}/bin/ipcrm -s $i
+                done
+            '';
+
+          serviceConfig = {
+            ExecStart = "${pkg}/bin/httpd -f ${httpdConf}";
+            ExecStop = "${pkg}/bin/httpd -f ${httpdConf} -k graceful-stop";
+            ExecReload = "${pkg}/bin/httpd -f ${httpdConf} -k graceful";
```

Comment:
~~It would be cool if the package could be overwritten.~~

Maybe don't name the option pkg but instead use cfg.package which is more obvious.

Reply:
the option itself is called `package` like it should be, `pkg` was copied from the existing implementation, I'm trying to keep the same style as the old one

Reply:
The existing module is very old and doesn't follow modern conventions. If you're entirely rewriting then feel free to abandon older style if out of date.